docs/dev/stream/state/broadcast_state.md

```diff
@@ -0,0 +1,279 @@
+                        .keyBy(new KeySelector<Shape, Color>(){...});
+{% endhighlight %}
+
+Moving on to the `Rules`, the stream containing them should be broadcasted to all downstream tasks, and these tasks 
+should store them locally so that they can evaluate them against all incoming `Items`. The snippet below will i) broadcast 
+the stream of rules and ii) using the provided `MapStateDescriptor`, it will create the broadcast state where the rules
+will be stored.
+
+{% highlight java %}
+
+// a map descriptor to store the name of the rule (string) and the rule itself.
+MapStateDescriptor<String, Rule> ruleStateDescriptor = new MapStateDescriptor<>(
+				"RulesBroadcastState",
+				BasicTypeInfo.STRING_TYPE_INFO,
+				TypeInformation.of(new TypeHint<Rule>() {})
+		);
+		
+// broadcast the rules and create the broadcast state
+BroadcastStream<Rule> ruleBroadcastStream = ruleStream
+                        .broadcast(ruleStateDescriptor);
+{% endhighlight %}
+
+Finally, in order to evaluate the `Rules` against the incoming elements from the `Item` stream, we need to:
+    1) connect the two streams and 
+    2) specify our match detecting logic. 
+
+Connecting a stream (keyed or non-keyed) with a `BroadcastStream` can be done by calling `connect()` on the 
+non-broadcasted stream, with the `BroadcastStream` as an argument. This will return a `BroadcastConnectedStream`, on 
+which we can call `process()` with a special type of `CoProcessFunction`. The function will contain our matching logic. 
+The exact type of the function depends on the type of the non-broadcasted stream: 
+ - if that is **keyed**, then the function is a `KeyedBroadcastProcessFunction`. 
+ - if it is **non-keyed**, the function is a `BroadcastProcessFunction`. 
+ 
+ Given that our non-broadcasted stream is keyed, the following snippet includes the above calls:
+
+<div class="alert alert-info">
+  <strong>Attention:</strong> The connect should be called on the non-broadcasted stream, with the `BroadcastStream`
+   as an argument.
+</div>
+
+{% highlight java %}
+DataStream<Match> output = colorPartitionedStream
+                 .connect(ruleBroadcastStream)
+                 .process(
+                     
+                     // type arguments in our KeyedBroadcastProcessFunction represent: 
+                     //   1. the key of the keyed stream
+                     //   2. the type of elements in the non-broadcast side
+                     //   3. the type of elements in the broadcast side
+                     //   4. the type of the result, here a string
+                     
+                     new KeyedBroadcastProcessFunction<Color, Item, Rule, String>() {
+                         // my matching logic
+                     }
+                 )
+{% endhighlight %}
+
+### BroadcastProcessFunction and KeyedBroadcastProcessFunction
+
+As in the case of a `CoProcessFunction`, these methods have two "sides", one is responsible for processing incoming 
```

Comment:
maybe we can just say -

As in the case of a `CoProcessFunction`, these functions have two process methods to implement; one is ......
IMO, its more direct, and more formal. What do you think?